streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java

```diff
@@ -194,11 +199,12 @@ public void computeTaskLags(final Map<TaskId, Long> allTaskEndOffsetSums) {
             final Long offsetSum = taskOffsetSums.getOrDefault(task, 0L);
 
             if (endOffsetSum < offsetSum) {
-                throw new IllegalStateException("Task " + task + " had endOffsetSum=" + endOffsetSum +
-                                                    " smaller than offsetSum=" + offsetSum);
-            }
-
-            if (offsetSum == Task.LATEST_OFFSET) {
+                LOG.warn("Task " + task + " had endOffsetSum=" + endOffsetSum +
+                             " smaller than offsetSum=" + offsetSum + ". This probably means the task is corrupted," +
+                             " which in turn indicates that it will need to restore from scratch, so we pin the lag" +
+                             " to the end offset of the log.");
+                taskLagTotals.put(task, endOffsetSum);
```

Comment:
This is the fix, to return an accurate estimate instead of throwing an exception. I felt a warning was appropriate, given that this does indicate task corruption, or some other unexpected situation.

Reply:
Should we report the lag as the whole log in this case? Even if the log is truncated it is not guaranteed to throw the invalid offset exception and hence task-corruption logic would not necessarily triggered.

Reply:
Do we know the exact reason, why the offset sum on the client is larger than the end offset sum on the broker?

Reply:
@guozhangwang if the end offset is less than the checkpointed offset, how is it possible to _not_ throw a `TaskCorruptedException`? I thought that was thrown after checking this exact condition?

Reply:
Also, @vvcephei , if a single task is corrupted and we do hit the `TaskCorruptedException` don't we close, wipe, and recreate stores of all tasks in the thread? Seems we would need to keep track of the thread ownership, and if we detect this condition then set the lag for all tasks owned by that thread to the `endOffset`

Reply:
A task is corrupted if the store changelogs in the checkpoint file don't match the store changelogs in the assigned task.

When the member sends its lag estimate, it doesn't validate whether the task is valid or corrupted. It just reads the values out of the checkpoint files and sums them. I think this is fine, actually, but if we want to check for task corrupted at that time, we can add that in.

I don't think it really changes how the assignor should behave, though. I think it's could easily happen that any past, present, or future versioned member reports a task sum that's beyond the end-offset sum of what the leader thinks the task contains. If that happens, it doesn't seem like it's a good idea to just crash the leader's thread. It would be better to handle it gracefully in some way.

It seems like assuming the task is corrupted, and therefore essentially that it isn't cached at all, is reasonable. This just de-prioritizes that member to get assigned the task, which is harmless.

Reply:
Whoops, edited my previous comment before refreshing & seeing your response. Makes sense, although I still believe we should be checking for this and throwing `TaskCorrupted` if we detect this case (in the owning StreamThread, not in the leader during assignment)

Reply:
> I think it's could easily happen that any past, present, or future versioned member reports a task sum that's beyond the end-offset sum of what the leader thinks the task contains.

If the corruption comes from our code, we should throw. The leader fetches the end offset shortly before that code, so they are most probably up-to-date. If the corruption is caused by failure we should handle it gracefully.

Reply:
Let's think about cases where this could possibly go wrong. We know it must reach or already be in standby/restoring phase. I will use `logEndOffset` for the actual `endOffset` of the partition, and `endOffset` for the offset a standby/restoring task is allowed to restore up to.

Newly-assigned: we read the invalid offset from the checkpoint, and determine the endOffset as either `logEndOffset` (normal) or `min(logEndOffset, LCO)` (optimized-source-changelog). In both cases, the `endOffset` <= `logEndOffset` so the invalid offset should also satisfy `offset` > `endOffset` which we can detect and throw as `TaskCorrupted`. 
Already-assigned: in this case, we must have been the ones who wrote the invalid offset and/or have it in our `checkpointableOffsets` map. But we still need to fetch the `endOffset` to know when to stop, so the above applies here as well and we can detect the corruption. 

In either case, I'd agree the leader should not throw and should just treat that task's offset sum as 0. But, we should make sure that on the other end we actually do detect this case and handle it the way the assignor expects.

Reply:
@ableegoldman If the log is truncated and then immediately more records are appended to go beyond the original log-end-offset, in old versions we would not throw the exception.

After some thoughts I think it makes sense to report the lag as the whole log.

Reply:
Thanks for the conversation, all. I think where it nets out for me is that we don't really expect this to happen for non-bugged code, but there are a few edge cases where it could. Thus, logging a warning actually seems reasonable.

Since the graceful handling option is still safe in any case, and since there do exist edge cases where this condition doesn't indicate a bug in Streams, we'll just keep the proposed graceful handling of reporting the lag as the whole log. Either the member in question won't get assigned the task (because it's de-prioritized wrt to other members who report valid positions), and it winds up cleaning up its state dir on its own, or it gets assigned the task, detects the corruption on its side, and recovers appropriately.